chain/client/src/client.rs

```diff
@@ -1661,4 +1669,119 @@ impl Client {
         //        }
         Ok(())
     }
+
+    fn record_receive_block_timestamp(&mut self, height: BlockHeight) {
+        if let Ok(tip) = self.chain.head() {
+            self.chunks_delay_tracker.add_block_timestamp(height, tip.height);
+        }
+    }
+    fn record_receive_chunk_timestamp(&mut self, height: BlockHeight, shard_id: ShardId) {
+        if let Ok(tip) = self.chain.head() {
+            self.chunks_delay_tracker.add_chunk_timestamp(height, shard_id, tip.height);
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct HeightInfo {
+    block_received: Option<Instant>,
+    chunks_received: HashMap<ShardId, Instant>,
+}
+
+#[derive(Debug, Default)]
+struct ChunksDelayTracker {
+    heights: BTreeMap<BlockHeight, HeightInfo>,
+    prev_head_height: BlockHeight,
+}
+
+const CHUNKS_DELAY_TRACKER_HORIZON: u64 = 10;
+
+impl ChunksDelayTracker {
+    fn lowest_height(head_height: BlockHeight) -> BlockHeight {
+        head_height.saturating_sub(CHUNKS_DELAY_TRACKER_HORIZON)
+    }
+    fn remove_old_entries(&mut self, head_height: BlockHeight) {
+        let heights_to_drop: Vec<BlockHeight> = self
+            .heights
+            .iter()
+            .take_while(|(&k, _)| k < ChunksDelayTracker::lowest_height(head_height))
+            .map(|(&k, _)| k)
+            .collect();
+        for h in heights_to_drop {
+            self.heights.remove(&h);
+        }
+    }
+
+    fn update_metrics(&mut self, head_height: BlockHeight) {
+        if head_height == self.prev_head_height {
+            return;
+        }
+        let mut cnt = 0;
+        let mut sum_delays = 0.;
+        for (&k, v) in &self.heights {
+            if k > head_height {
+                break;
+            }
```

Comment:
If I understand correctly, every time when `update_metrics` is called, it prints stats on block with height 
 between [head_height - CHUNKS_DELAY_TRACKER_HORIZON, head_height]. Does that mean delay on the same block can be printed more than twice?

Additionally, in the case of forks, a block at `height+1` can be processed before a block at `height`, if they are on different forks of the chain. So blocks with height <= head_height are not necessarily processed.

Therefore, I think it's the best to update metrics on a specific block by specifically checking that it has finished processing and remove it from the map. One way to check if a block is processed successfully is check `Chain::get_block_by_height` returns successfully.

Reply:
I agree that from the correctness point of view your approach is much better.
I want to build a prometheus metric, it's limited to being polled every 5-10 seconds. To provide a clear enough picture with that sampling rate I suggest to average metrics of the previous few blocks.
Forks is an important problem, but it should be ok for a monitoring utility not to worry about it.